Check for pip configuration when an externally managed environment is detected.

nuclearsandwich opened the pull request:
rosdep is designed to treat pip like an alternative system-level package manager. Deviating from this approach is not easily achievable without a significant rethinking of how pip packages are managed. In the meantime, we can at least instruct users how to restore the prior functionality.

Rather than inject the environment variable / config on behalf of the user, this change instructs them to make the necessary config changes themself, keeping them informed of the change their making to the system's new default.

I added a documentation page specific to this issue with additional configuration details and recipes for CI.

Seeks to reduce friction related to #978

nuclearsandwich committed:
Raise an error when an externally managed environment is detected.

rosdep is designed to treat pip like an alternative system-level package
manager. Deviating from this approach is not easily achievable without a
significant rethinking of how pip packages are managed. In the meantime,
we can at least instruct users how to restore the prior functionality.

Rather than inject the environment variable / config on behalf of the
user, this change instructs them to make the necessary config changes
themself, keeping them informed of the change their making to the
system's new default.

nuclearsandwich commented:
In addition to the test failures, this actually doesn't work as intended because `sudo` is stripping the environment. I'm hesitant to recommend the pip global config file as a default because the default location is variable. Alternatively, we could override the sudo_command for the pip installer to pass that environment variable through.

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/rosdep-for-pip-is-broken-on-jazzy/38981/5

christophebedard commented on src/rosdep2/platforms/pip.py:
```diff
@@ -43,6 +46,15 @@
 PIP_INSTALLER = 'pip'
 
+EXTERNALLY_MANAGED_EXPLAINER = """
+rosdep installation of pip packages requires installing packages packages globally as root
```
```suggestion
rosdep installation of pip packages requires installing packages globally as root.
```

christophebedard commented on src/rosdep2/platforms/pip.py:
```diff
@@ -70,6 +82,35 @@ def get_pip_command():
+    if sys.version_info[0] == 3 and sys.version_info[1] >= 11:
+        if "PIP_BREAK_SYSTEM_PACKAGES" in os.environ and os.environ[
+            "PIP_BREAK_SYSTEM_PACKAGES"
+        ].lower() in ["yes", "1", "true"]:
```
```suggestion
        ].lower() in ("yes", "1", "true"):
```

nuclearsandwich committed:
Fix typos and punctuation.

Co-authored-by: Christophe Bedard <bedard.christophe@gmail.com>

nuclearsandwich committed:
Use tuple rather than list literal.

Co-authored-by: Christophe Bedard <bedard.christophe@gmail.com>

nuclearsandwich committed:
Consolidate version check.

nuclearsandwich committed:
Pass necessary environment variable via sudo.

nuclearsandwich committed:
Update test to expect sudo --preserve-env for pip.

nuclearsandwich committed:
flake8 cleanup

nuclearsandwich commented:
> In addition to the test failures, this actually doesn't work as intended because `sudo` is stripping the environment. I'm hesitant to recommend the pip global config file as a default because the default location is variable. Alternatively, we could override the sudo_command for the pip installer to pass that environment variable through.

As of https://github.com/ros-infrastructure/rosdep/pull/979/commits/2c76686afa24ff09f52642a4b72116c2051ae5e3 the installer is configured to pass `PIP_BREAK_SYSTEM_PACKAGES` through from the user environment. This works by default on each of my tested sudo environments (Archlinux, Ubuntu, Fedora) but a locked down sudo that disallows environment preservation or has an explicit allowlist will still create problems. My hope was to detect that the check passed but the install failed due to the externally managed target anyway and give a hint about sudo to suggest an alternative configuration method. But without changes, the pip installer never gets control back from `install_resolved()` so we'd have to add pip-specific hooks elsewhere in the code and I'm hesitant. I think I'm content to leave it here but with more robust documentation.

nuclearsandwich committed:
Run pip tests with PIP_BREAK_SYSTEM_PACKAGES=1.

nuclearsandwich committed:
Add documentation for pip configuration.

nuclearsandwich committed:
Add doc link to error output.

nuclearsandwich committed:
Use inline monospace font to refer to rosdep the cli tool.

nuclearsandwich committed:
Briefly note that sudo configuration could prevent this from working.

nuclearsandwich committed:
Recommend a specific config file to use and format user config as a warning.

nuclearsandwich committed:
Fix errors in config checker.

The fallback configuration was over-indented and would never be
checked.

nuclearsandwich commented on src/rosdep2/platforms/pip.py:
```diff
@@ -43,6 +46,17 @@
+PIP_BREAK_SYSTEM_PACKAGES=1
+in your environment.
+
+For more information refer to http://docs.ros.org/en/independent/api/rosdep/html/pip_and_pep_668.html
```
Note that this documentation won't actually be live until the PR is merged and docs rebuild the next day. This will only affect those using rosdep from mainline or this branch. I'll wait until the docs rebuild to make a new rosdep release with this change.

christophebedard commented on doc/pip_and_pep_668.rst:
```diff
@@ -0,0 +1,72 @@
+The string values are not case sensitive.
+
+``rosdep`` is designed to use ``sudo`` in order to gain root privileges for installation when not run as root.
+If your system's sudo configuration prohibits the passing of environment variables
```
it looks like this sentence was cut off.

nuclearsandwich replied:
lol good catch. I'm sure that I went to look up how to reference a section in the doc below and got distracted because that's exactly what I just did again.

christophebedard commented on doc/pip_and_pep_668.rst:
```diff
@@ -0,0 +1,72 @@
+Configure using pip.conf
+------------------------
+
+`Pip configuration files <pip-configuration>`_ can be used to set the desired behavior 
```
```suggestion
`Pip configuration files <pip-configuration>`_ can be used to set the desired behavior.
```

christophebedard commented on doc/pip_and_pep_668.rst:
```diff
@@ -0,0 +1,72 @@
+------------------------
+
+`Pip configuration files <pip-configuration>`_ can be used to set the desired behavior 
+Pip checks for global configuration files in ``XDG_CONFIG_DIRS``, as well as ``/etc/pip.conf``
```
```suggestion
Pip checks for global configuration files in ``XDG_CONFIG_DIRS``, as well as ``/etc/pip.conf``.
```

christophebedard commented on doc/pip_and_pep_668.rst:
```diff
@@ -0,0 +1,72 @@
+
+`PEP-668`_ introduced `externally managed environments <externally-managed-environments>`_ to Python packaging.
+
+``rosdep`` is designed to use pip as an alternative system package manager, rosdep installation of pip packages requires installing packages globally as root.
```
this document is a bit inconsistent with rosdep and `rosdep`, but I'll let you decide if you want to change anything.

nuclearsandwich replied:
My mental style guide is `rosdep` when referring to the CLI tool and rosdep when referring to the project. I am always tempted into using `rosdep` in appropriately when its the first word of the sentence because Rosdep and ROSdep are both wrong.

I will reword to avoid the entire problem :D (that mainly exists in my head)

nuclearsandwich replied:
Actually, I'm just going to get over myself and start the sentence with the proper lowercase noun. It's the 2020s.

christophebedard commented on src/rosdep2/platforms/pip.py:
```diff
@@ -70,6 +84,41 @@ def get_pip_command():
+def externally_managed_installable():
+    """
+    PEP 668 enacted in Python 3.11 blocks pip from working in "externally
+    managed" envrionments such operating systems with included package
```
```suggestion
    managed" environments such as operating systems with included package
```

christophebedard commented on src/rosdep2/platforms/pip.py:
```diff
@@ -70,6 +84,41 @@ def get_pip_command():
+    variable.
+    """
+    if sys.version_info >= (3, 11):
+        if 'PIP_BREAK_SYSTEM_PACKAGES' in os.environ and os.environ[
+            'PIP_BREAK_SYSTEM_PACKAGES'
+        ].lower() in ('yes', '1', 'true'):
```
minor suggestion, but, if you want, you could improve readability a little bit by changing it to something like:

```suggestion
        if (
            'PIP_BREAK_SYSTEM_PACKAGES' in os.environ
            and os.environ['PIP_BREAK_SYSTEM_PACKAGES'].lower() in ('yes', '1', 'true')
        ):
```

not actually sure if exact change passes linters, though.

nuclearsandwich replied:
The only think flake8 got upset by was the and being on the continuing line rather than the end of the first. I like this form better thanks.

nuclearsandwich replied:
This suggestion was incorporated into a2e9a1e

christophebedard replied:
I think you have an extra indentation level there:

```diff
     if (
-             'PIP_BREAK_SYSTEM_PACKAGES' in os.environ and
-             os.environ['PIP_BREAK_SYSTEM_PACKAGES'].lower() in ('yes', '1', 'true')
+         'PIP_BREAK_SYSTEM_PACKAGES' in os.environ and
+         os.environ['PIP_BREAK_SYSTEM_PACKAGES'].lower() in ('yes', '1', 'true')
     ):
         return True
```

in my experience the usual linters don't complain about this for some reason.

> The only think flake8 got upset by was the and being on the continuing line rather than the end of the first.

Ah yeah I always get bit by that!

christophebedard commented on src/rosdep2/platforms/pip.py:
```diff
@@ -70,6 +84,41 @@ def get_pip_command():
+    flagrantly named "break system packages" config option or environment
+    variable.
+    """
+    if sys.version_info >= (3, 11):
```
you could flip this to return early and remove one indent level from the rest of the function

```suggestion
    if sys.version_info < (3, 11):
        return True
```

christophebedard commented on test/test_rosdep_pip.py:
```diff
@@ -67,6 +67,7 @@ def test_PipInstaller_get_depends():
     assert ['foo'] == installer.get_depends(dict(depends=['foo']))
 
 
+@patch.dict(os.environ, {'PIP_BREAK_SYSTEM_PACKAGES': '1'})
```
would be nice to test that the command fails if this isn't set and the version is >= 3.11, but you'd also need to make sure that the other config methods are not used, or just patch `externally_managed_installable()`, but I would leave that up to you.

nuclearsandwich replied:
I'm limited somewhat by my lack of experience with Python's mocking. Let me see if I can skim enough examples and read enough of the docs to stub the method for a few tests before.

It's a good push. Tests added in 8b48f13

christophebedard reviewed (commented):
Minor comments!

nuclearsandwich committed:
Change formatting of rosdep.

Use monospace formatting when referring to the `rosdep` command /
executable name and simply 'rosdep' when referring to the project.

The preferred capitalization of rosdep is rosdep not Rosdep or ROSdep
(and certainly not ROSDep).

nuclearsandwich committed:
Complete a sentence I stopped writing.

nuclearsandwich committed:
Add period to end of sentence.

Co-authored-by: Christophe Bedard <bedard.christophe@gmail.com>

nuclearsandwich committed:
Add period to end of sentence.

Co-authored-by: Christophe Bedard <bedard.christophe@gmail.com>

nuclearsandwich committed:
Invert conditional for an earlier return.

Co-authored-by: Christophe Bedard <bedard.christophe@gmail.com>

nuclearsandwich committed:
Edit text for clarity and typos.

Co-authored-by: Christophe Bedard <bedard.christophe@gmail.com>

nuclearsandwich committed:
Reflow conditional for easier reading.

nuclearsandwich committed:
Fix control flow after inverting the conditional.

This is a fixup after two earlier changes inverted the conditional and
reformatted an internal check.

nuclearsandwich committed:
Add test to confirm that get_install_command handles externally managed environments.

nuclearsandwich commented on test/test_rosdep_pip.py:
```diff
@@ -67,6 +67,23 @@ def test_PipInstaller_get_depends():
 
+@patch('rosdep2.platforms.pip.externally_managed_installable')
+def test_PipInstaller_handles_externally_managed_environment(externally_managed_installable):
+    from rosdep2 import InstallFailed
+    from rosdep2.platforms.pip import EXTERNALLY_MANAGED_EXPLAINER, PipInstaller
```
I'm not super up on why imports are done per test function here but I am propagating the convention until I understand it better.

christophebedard commented on test/test_rosdep_pip.py:
```diff
@@ -67,6 +67,23 @@ def test_PipInstaller_get_depends():
+
+    externally_managed_installable.return_value = False
+    installer = PipInstaller()
+    try:
+        installer.get_install_command(['whatever'])
+        assert False, 'should have raised'
+    except InstallFailed as e:
+        assert e.failures == [('pip', EXTERNALLY_MANAGED_EXPLAINER)]
```
If you want, a more Pythonic way could be:

```suggestion
    with pytest.raises(InstallFailed) as exception_info:
        installer.get_install_command(['whatever'])
    assert exception_info.value.failures == [('pip', EXTERNALLY_MANAGED_EXPLAINER)]
```

and make sure to `import pytest`

nuclearsandwich replied:
Thanks for the suggestion. I'm actually going to keep this as is for now, because I don't believe the infrastructure packages ever explicitly require pytest. I think pytest is supposed to be compatible with stdlib unittest as well and utilizing that may be the preferred approach.

christophebedard reviewed (approved):
Aside from my last minor suggestions, this looks good to me!

christophebedard commented on src/rosdep2/platforms/pip.py:
```diff
@@ -70,6 +84,45 @@ def get_pip_command():
+
+    # Check the same configuration directories as pip does per
+    # https://pip.pypa.io/en/stable/topics/configuration/
+    pip_config = ConfigParser()
+    if 'XDG_CONFIG_DIRS' in os.environ:
+        for xdg_dir in os.environ['XDG_CONFIG_DIRS'].split(':'):
+            pip_config_file = Path(xdg_dir) / 'pip' / 'pip.conf'
+            pip_config.read(pip_config_file)
+            if pip_config['install']['break-system-packages']:
+                return True
+
+    fallback_config = Path('/etc/pip.conf')
+    pip_config.read(fallback_config)
+    if pip_config['install']['break-system-packages']:
```
I gave it a try and it seems like, if:

1. the config file doesn't exist; or
2. the `[install]` section doesn't exist; or
3. a `break-system-package` value isn't specified.

then `pip_config[...]` raises a `KeyError`. Looks like you could avoid checking all of that by using `getboolean()`:

```py
pip_config.getboolean('install', 'break-system-packages', fallback=False)
```

nuclearsandwich replied:
Good catch!

nuclearsandwich replied:
Resolved in [8e94ef1](https://github.com/ros-infrastructure/rosdep/pull/979/commits/8e94ef1ffb01394b23cea9dcc12fbb6fc92f10d2)

nuclearsandwich committed:
Use ConfigParser.getboolean to check config value.

Using the dict access method will raise a KeyError when the config file
is present but this section or value is missing.
